### PR TITLE
[#368] Add Advanced Options Step to Plan Wizard

### LINKED ISSUE #368: Add Pre & Post Playbook Support to Migration Plans - Plan Wizard Changes
The user should be able to identify a pre and/or post Ansible Playbook as part of a migration plan.

Playbook Support Notes: https://docs.google.com/document/d/1dnb5RsMvVRRx4osTIeR6wTF9JXQwoNQNwW-pZsW65Ao/edit#

### PR DESCRIPTION
**Pull request by michaelkro**
fixes #368 

## TODO
- [x] Fetch playbooks via API
- [x] Handle loading state when fetching playbooks
- [x] Handle error state if fetching playbooks fails
  * **Will handle in separate issue** [(comment)](https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/413#pullrequestreview-130917033)
  * [Issue](https://github.com/ManageIQ/miq_v2v_ui_plugin/issues/463)
- [x] Modify plan POST to include playbooks data
- [x] ~Add link to where user can create a ansible playbook catalog item (new UX request)~
  **Notes**: 
  1. The creation form does not have its own URL, so it's not possible to link directly to it.
  2. The best we can do is to link to `Services -> Catalogs`, and so, it may be difficult for a new CF user to figure out what to do next.
  3.  **Won't do**
- [x] Confirmation dialog for playbook(s) selected with no VMs
- [x] Clear checkboxes if user clears a playbook
- [x] Clear all form data if the user moves back to Step 2.
  **Reasoning**:  If the user goes back to Step 2 and adds/removes a vm(s), things might get weird.  The form data will persist, however, if the user moves forward to the Scheduling step and then comes back

## Questions
1. Do we want to display anything in particular if fetching playbooks returns nothing?
2. What do we want to display if fetching playbooks fails?
3. Is there any way to differentiate migration specific playbooks from the rest?

## Notes
* ~Stub actions/reducer for fetching playbooks~
* Copy and modify VMsStep table
* ~Copy and modify BootstrapSelect~
* Modify and reuse `BootstrapSelect` for Advanced Options use case
* Rename Options Step to Schedule Step
* ~If the Advanced Options Step is revisited, i.e., the user navigates forward or backwards from the step and then back again, all form data will be cleared.~

  **Reasoning**:  ~Data regarding VM selection lives in the table's component state, and thus, is wiped when leaving the step (forward or backward).  To persist, this state should be lifted to redux store.~

  **Update**:  Form data persists when moving forward to the Scheduling Step.  i.e., when navigating back to the Advanced Options Step, playbook(s) and VM selections will have been preserved. If the user navigates backwards, from the Advanced Options Step to the VMs Step, all form data will be cleared.  i.e., if the user goes back to the VM selection step, and then forward again to the Advanced Options Step, the dropdowns and table will have been returned to their initial states

* Should be merged alongside https://github.com/ManageIQ/manageiq/pull/17627.  For now, plan creation will fail in this branch
  **Update**:  Corresponding back end PRs have been merged

### No Playbooks Selected
fullscreen_6_15_18__7_18_am

### Top Dropdown Open
fullscreen_6_15_18__7_26_am

### One Playbook Selected
advanced-options-3

### Selected VMs Totals
advanced-optiones-4

### All Checkboxes Enabled
advanced-options-5

### Confirm Dialog
fullscreen_6_21_18__1_44_pm

### Redux Form State
fullscreen_6_14_18__12_32_pm

### POST Body
devtools_-_localhost_8080_migration

### CODE REVIEW

**Comment by vconzola** [comment]
@michaelkro this is great work. To answer your questions...
1. No, we just leave none selected as the only option in the dropdown.
2. I'm thinking we display an in-line error. I believe notifications are displayed only as the result of a user action, but let me double check with Mgmt UX team. Then I can mock something up.

VM selection table should be shown with both columns disabled if no playbooks are selected.

**Comment by priley86** [comment]
@agrare @AparnaKarve can you possibly provide some API feedback for us w.r.t. to @michaelkro's questions above? We will need to query the available pre/post migration playbooks to drive the dropdowns in the plan wizard and then provide the selected pre/post somewhere in the service template POST.

Additionally, we will need to ensure that the tasks now indicate pre/post playbook status when visualizing on the details page. We should be able to drive the Overview cards (aggregates) by inspecting the individual tasks. The full UX is here:
https://docs.google.com/presentation/d/19CpXErM4mDsbLkx6ahQMpDn_i63eqb805GwM4M3Kl1Q/edit#slide=id.g3c2c648750_0_38

Can you give us an update on these APIs?

**Comment by agrare** [comment]
@priley86 I think we need to meet to discuss how we plan on doing this because neither @fdupont-redhat nor I know how this will look in the backend.

**Comment by michaelkro** [comment]
@vconzola unfortunately, we can't close this until we hook it up to the back end, otherwise it'll break the Plan Wizard

I'm almost done putting the final UI touches though.  If we have time, I'll demo what I have in standup this afternoon

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardVMStep/components/PlanWizardVMStepTable.js [hunk 1]
```diff
@@ -21,9 +21,9 @@ import {
   PAGINATION_VIEW
 } from 'patternfly-react';
 
-import rowFilter from './rowFilter';
-import searchFilter from './searchFilter';
-import CustomToolbarFind from './CustomToolbarFind';
+import rowFilter from '../../common/rowFilter';
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardVMStep/components/PlanWizardVMStepTable.js]: excellent! wonderful hoist here 👍 

though maybe the sortFilter listFilter or paginate could be as well?  Unclear if these can be used in this work.. will look at deeper on pass 2

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardAdvancedOptionsStep/components/PlanWizardAdvancedOptionsStepTable/PlanWizardAdvancedOptionsStepTable.js [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+      this.setState({ activeFilters: updated });
+    }
+  };
+
+  updateCurrentValue = event => {
+    this.setState({ currentValue: event.target.value });
+  };
+
+  filterAdded = (field, value) => {
+    let filterText = field.title;
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardAdvancedOptionsStep/components/PlanWizardAdvancedOptionsStepTable/PlanWizardAdvancedOptionsStepTable.js]: template literal? unclear whyyyyyyyyy were building a string 🤔

**Comment by priley86** [comment]
this is looking really good to me - only thoughts/questions so far:

1. do we plan to support pre/post playbook in G branch? It seems the answer is yes. Only reason I ask is, we should be able to pull in the `MessageDialog` from upstream PF React in `master` now (PF was updated there). I think this can be a follow up item...
2. We are starting to repeat more of the table code. I think this is OK for now since we are still trying to understand what is common and can be abstracted. My vote would be to create a follow up issue for abstracting some of the common table logic into a base es6 class (personally I prefer this over Recompose to avoid the additional dependency and challenges testing). We could experiment w/ this and see what works best...

Would be nice to review this once more once API has been integrated. 

thanks @michaelkro !

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.scss [hunk 3]
```diff
@@ -0,0 +1,9 @@
+@import './components/PlanWizardAdvancedOptionsStep/PlanWizardAdvancedOptionsStep.scss';
+
+.plan-wizard-confirm-backdrop {
+  z-index: 1050;
```
[the remark this answers is not in the document]
Comment (michaelkro) [reply app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.scss]: Ya, I'm not sure how else to handle this...  Def open to solutions!

**Comment by AllenBW** [comment]
@michaelkro so this is good for final review right? Cant for the life of me figure out.. why we haven't... (but i might be missing something)

**Comment by michaelkro** [comment]
@AllenBW I think we are ready :)

**Comment by priley86** [comment]
@michaelkro @mturley any update on #400  / #399 ? I think it might be nice to merge those alongside this one... otherwise the user has no indication what's happening?

I will go ahead and try to test this out again locally now though...

**Comment by AllenBW** [comment]
If yah wanta wait to merge this with other issues, gonna label it appropriately *BAM*

**Comment by priley86** [comment]
I was able to test this using the `6_0_dc_MBU_demo_01202018.backup` extract. Looks good and it seems you have handled all the ux cases! Nice job @michaelkro !

screen shot 2018-07-11 at 11 49 37 am

As mentioned earlier in the PR, I went ahead and filed [483](https://github.com/ManageIQ/miq_v2v_ui_plugin/issues/483) for future consolidation of PF React Tables. 

Will go ahead and approve, but prefer to wait on merge for #400 and #399 still...

**Comment by AparnaKarve** [comment]
Tested this on an appliance and it's looking really great so far.

Thanks @michaelkro ! Nice job!!

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.js [hunk 4]
```diff
@@ -42,22 +53,48 @@ class PlanWizard extends React.Component {
...
+      const onConfirm = () => {
+        this.setState({ activeStepIndex: 3 });
+        hideConfirmModalAction();
+      };
+
+      showConfirmModalAction({
+        title: __('No VMs Selected'),
+        body: __(
+          "You've selected a pre-migration or post-migration playbook service but no VMs on which to run the playbook service. Are you sure you want to continue?"
+        ),
```
Comment (AparnaKarve) [inline app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.js]: Please make sure that the entire gettext text appears in a single.
And then mark that line with `// prettier-ignore`

Comment (michaelkro) [reply app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizard.js]: whoops!

**Comment by simaishi** [comment]
Backported to Gaprindashvili via https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/518

**Pull request merged**